File: backend/src/models/message.model.js
import messageSchema from "../lib/schema/message-schema.js";
import logger from "../helpers/logger.js";

async function insert({ message, sender, receiver, type, isRead, time }) {
  try {
    return await messageSchema.create({
      message,
      sender,
      receiver,
      type,
      isRead,
      time,
    });
  } catch (err) {
    logger.error(`MessageModel -> insert: ${err.message}`);
    throw err;
  }
}

async function deleteByReceiver({ to }) {
  try {
    return await messageSchema.deleteMany({ to });
  } catch (err) {
    logger.error(`MessageModel -> deleteByReceiver: ${err.message}`);
    throw err;
  }
}

async function getByReceiver({ to }) {
  try {
    return await messageSchema.find({ to }).lean();
  } catch (err) {
    logger.error(`MessageModel -> getByReceiver: ${err.message}`);
    throw err;
  }
}

export default { insert, deleteByReceiver, getByReceiver };

File: frontend/lib/data/models/message_model.dart
class MessageModel {
  final String? id;
  final String message;
  final String sender;
  final String receiver;
  final String? type;
  final bool isRead;
  final String time;

  MessageModel({
    this.id,
    required this.message,
    required this.sender,
    required this.receiver,
    this.type,
    required this.isRead,
    required this.time,
  });

  factory MessageModel.fromJson(Map<String, dynamic> json) {
    return MessageModel(
      id: json['id'] != null ? json['id'] as String : null,
      message: json['message'] as String,
      sender: json['sender'] as String,
      receiver: json['receiver'] as String,
      type: json['type'] as String,
      isRead: json['isRead'] as bool,
      time: json['time'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'message': message,
      'sender': sender,
      'receiver': receiver,
      'type': type ?? 'text',
      'isRead': isRead==true? 1 : 0,
      'time': time,
    };
  }
}
